docs(SDB): add ReactJS x Serverless DB/Containers tutorial MTA-4243 (#2824)
scaleway/docs-content: tutorials/nextjs-app-serverless-functions-sqldb/index.mdx

---
meta:
  title: Deploy a Next.js application using Serverless Containers and Serverless SQL Databases

>   description: This page explains how to deploy your fully serverless Next.js application using Serverless Containers and Serverless SQL Databases

Reviewer: ```suggestion
  description: This page explains how to deploy your serverless Next.js application entirely, using Serverless Containers and Serverless SQL Databases
```

Reviewer: I'm not sure I'm familiar with what "fully" means in this context, feel free to ignore if my suggestion is irrelevant

content:
  h1: Deploy a Next.js application using Serverless Containers and Serverless SQL Databases
  paragraph: This page explains how to deploy your fully serverless Next.js application using Serverless Containers and Serverless SQL Database
dates:
  validation: 2024-02-29

>   posted: 2024-02-29

Reviewer 2: Maybe change this to today?

---

This tutorial will guide you through the deployment of a fully serverless Next.js application using a Serverless Container and a Serverless SQL Database in a few minutes.

You can either deploy your application:
  - [step by step using Scaleway CLI](#deploy-next.js-application-frontend-using-serverless-containers) to understand each detailed action performed, and the resources and subresources required
  - [using a Terraform template](#deploy-next.js-application-using-terraform-templates) to deploy your application faster and have a ready-to-use Infrastructure as Code.

> <Macro id="iam-requirements" />

Reviewer 3: ```suggestion
<Macro id="requirements" />
```

> <Message type="requirement">

Reviewer 3: ```suggestion
```

>   - You have an account and are logged into the [Scaleway console](https://console.scaleway.com/)

Reviewer 3: ```suggestion
- A Scaleway account logged into the [console](https://console.scaleway.com)
```

>   - You have installed and initialized the [Scaleway CLI](https://www.scaleway.com/en/docs/developer-tools/scaleway-cli/quickstart/)

Reviewer 2: ```suggestion
  - You have installed and initialized the [Scaleway CLI](/developer-tools/scaleway-cli/quickstart/)
```

>   - You have installed [Docker Engine](https://docs.docker.com/engine/install/)

Reviewer 3: ```suggestion
 - A valid [API key](/identity-and-access-management/iam/how-to/create-api-keys/)
  - Installed [Docker Engine](https://docs.docker.com/engine/install/)
```

>   - You have installed [Node 18 or newer](https://nodejs.org/en/download)

Reviewer 3: ```suggestion
- Installed [Node 18 or newer](https://nodejs.org/en/download)
```

> </Message>

Reviewer: I think you should add the having created an API key requirement, in the requirements section

Reviewer 3: @nerda-codes added it, since I was already reviewing it :)

## Deploy Next.js application frontend using Serverless Containers

### Initializing the project

> 1. Run the command below in a terminal to export your access key and secret key as environment variables:

Reviewer: ```suggestion
1. Run the command below in a terminal to export your API key's access and secret keys as environment variables:
```

    ```sh
    export SCW_ACCESS_KEY=$(scw config get access-key)
    export SCW_SECRET_KEY=$(scw config get secret-key)
    ```

2. Run the command below to make sure the environment variables are properly set:

    ```sh
    scw info 
    ```

>     This command displays your access key and secret key as the last two lines. The `ORIGIN` column displays `env (SCW_ACCESS_KEY)` and `env (SCW_SECRET_KEY)`, and not `default profile`.

Reviewer: ```suggestion
    This command displays your access key and secret key in the last two lines of the output. The `ORIGIN` column should display `env (SCW_ACCESS_KEY)` and `env (SCW_SECRET_KEY)`, and not `default profile`.
```

      
    ```bash
    KEY                      VALUE                                       ORIGIN
    (...)
    access_key               SCWF9DETBF829TAFB1TB                        env (SCW_ACCESS_KEY)
    secret_key               9a1bcf92-xxxx-xxxx-xxxx-xxxxxxxxxxxx        env (SCW_SECRET_KEY)
    ```

3. Access the folder where you want to store your project repository and run the following command to initialize a blog template application:
    ```bash
    npx create-next-app --example blog-starter my-nextjs-blog
    ```

4. Type `y` when prompted to install the dependencies:
    ```bash
    Need to install the following packages:
    create-next-app@14.0.3
    Ok to proceed? (y) y
    ```
    This will create a folder named `my-nextjs-blog`. 
    
5. Access the folder you just created:
    ```bash
    cd my-nextjs-blog
    ```

6. Make sure you can run the application locally:
    ```bash
    npm run dev
    ```

> 7. Go to [localhost:3000](http://localhost:3000) in a browser. The blog template displays.

Reviewer 2: ```suggestion
7. Go to `http://localhost:3000` in a browser. The blog template displays.
```

iirc if there is a clickable link that leads to nowhere it will trigger a 404 error when we run our script

    ```bash
    touch Dockerfile
    ```
2. Add the following code to it:
    ```bash
    # syntax=docker/dockerfile:1

    FROM node:20-alpine
      
    #Built time arguments used for pre-rendering
    ARG PGHOST
    ARG PGPORT
    ARG PGDATABASE
    ARG PGUSER
    ARG PGPASSWORD

    WORKDIR /usr/app
    COPY ./ ./
    RUN npm install
    RUN npm run build
      
    #Web application configuration
    ENV PORT=8080
      
    #Database configuration used for dynamically rendered data. These default values should be overwritten by container runtime environment variables
    ENV PGHOST=localhost
    ENV PGPORT=5432
    ENV PGDATABASE=database
    ENV PGUSER=user
    ENV PGPASSWORD=password
      
    CMD npm run start
    ```

3. Create a `.dockerignore` file by copying the `.gitignore` file:
    ```bash
    cp .gitignore .dockerignore
    ```

    Docker will now ignore node dependencies and build files generated during the container build.

4. Build your application container:
    ```bash
    docker build -t my-nextjs-blog .   
    ```

5. Check that your container runs locally:
    ```bash
    docker run -it -p 8080:8080 my-nextjs-blog
    ```

> 6. Go to [localhots:8080](http://localhost:8080) in your browser. The blog template displays.

Reviewer 2: ```suggestion
6. Go to `http://localhost:8080` in your browser. The blog template displays.
```
same as above

> 6. Go to [localhots:8080](http://localhost:8080) in your browser. The blog template displays.

Reviewer: ```suggestion
6. Copy and paste `http://localhost:3000`  in a browser. The blog template displays.
```

    <Message type="tip">

>     When connecting to the webpage, your terminal running docker might display the following error `Error: ENOENT: no such file or directory, open '/usr/app/_posts/%5Bslug%5D.md'`. This will be fixed in further steps.

Reviewer 2: ```suggestion
    When connecting to the webpage, your terminal running Docker might display the following error `Error: ENOENT: no such file or directory, open '/usr/app/_posts/%5Bslug%5D.md'`. This will be fixed in further steps.
```

    </Message>

### Pushing the image to Scaleway Container Registry

> 1. Run the following command to create a [Container Registry namespace](https://www.scaleway.com/en/docs/containers/container-registry/concepts/#namespace) and export its endpoint as a variable:

Reviewer 2: ```suggestion
1. Run the following command to create a [Container Registry namespace](/containers/container-registry/concepts/#namespace) and export its endpoint as a variable:
```

    ```bash
    export REGISTRY_ENDPOINT=$(scw registry namespace create -o json | jq -r '.endpoint')
    ```
2. Log in to your [Container Registry](/containers/container-registry/concepts/#container-registry) from your local terminal:
    ```bash
    docker login $REGISTRY_ENDPOINT -u nologin --password-stdin <<< "$SCW_SECRET_KEY"
    ```
    The following output displays:
    ```sh
    Login Succeeded
    ```

3. Tag and push your container image to your Container Registry namespace:
    ```bash
    docker tag my-nextjs-blog:latest $REGISTRY_ENDPOINT/my-nextjs-blog:latest
    docker push $REGISTRY_ENDPOINT/my-nextjs-blog:latest
    ```

### Deploying the container with Scaleway Serverless Containers

> 1. Run the following command to create a [Serverless Containers namespace](https://www.scaleway.com/en/docs/serverless/containers/concepts/#namespace) and export its ID as a variable:

Reviewer 2: ```suggestion
1. Run the following command to create a [Serverless Containers namespace](/serverless/containers/concepts/#namespace) and export its ID as a variable:
```

    ```bash
    export CONTAINER_NAMESPACE_ID=$(scw container namespace create name="my-nextjs-blog-ns" -o json | jq -r '.id')
    ```

2. Deploy a Serverless Container with your application image:
    ```bash
    scw container container create name="my-nextjs-blog" namespace-id=$CONTAINER_NAMESPACE_ID registry-image=$REGISTRY_ENDPOINT/my-nextjs-blog:latest
    ```

    <Message type="tip">
    The first deployment can take up to two minutes. You can check the deployment status with the following command:
    ```sh
    scw container container list name=my-nextjs-blog
    ```
    When the status appears as `ready`, you can access the website via your browser.
    </Message>
  
3. Copy the endpoint URL displayed next to the `DomainName` property, and paste it into your browser. The blog template displays.

Congratulations, your Next.js application frontend is now online!

> Your application cannot store any data persistently yet, we will now add a serverless storage in the next section.

Reviewer 3: ```suggestion
Your application cannot store any data persistently yet. We will add a serverless storage in the next section.
```

## Add serverless storage to your application using Serverless SQL Database

### Setting up the database

> 1. Run the following command to create a [Serverless SQL Database](https://www.scaleway.com/en/docs/serverless/sql-databases/reference-content/serverless-sql-databases-overview/) and export its attributes as variables:

Reviewer 2: ```suggestion
1. Run the following command to create a [Serverless SQL Database](/serverless/sql-databases/reference-content/serverless-sql-databases-overview/) and export its attributes as variables:
```

     '/assets/blog/preview/cover.jpg',
     '2020-03-16T05:35:07.322Z',
     'Joe Haddad',
     '/assets/blog/authors/joe.jpeg',
     '/assets/blog/preview/cover.jpg',
     'preview',
     'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Praesent elementum facilisis leo vel fringilla est ullamcorper eget. At imperdiet dui accumsan sit amet nulla facilities morbi tempus. Praesent elementum facilisis leo vel fringilla. Congue mauris rhoncus aenean vel. Egestas sed tempus urna et pharetra pharetra massa massa ultricies.');
   ```

5. Exit the database query interface with the following command:
    ```bash
    exit
    ```

6. Run the command below to add the `node-postgres` dependency to your application:
    ```bash
    npm install pg && npm install --save-dev @types/pg
    ```

    This library is required to connect from your front end to your database.

> Your application folder should be called `my-nextjs-blog` and structured such as follows:

Reviewer 3: ```suggestion
Your application folder should be called `my-nextjs-blog` and structured as follows:
```

  ```bash
  my-nextjs-blog/
      .next/ 
      app/
      node_modules/
      public/
      pages/
      src/
      package.json
      tsconfig.json
  ```

### Integrating the database to your application

You need to edit 3 files to have a proper integration with your database.

<Message type="tip">

> If you want to avoid doing these modifications manually, you can also directly clone the final code from this tutorial from a remote repository with `git clone git@gitlab.infra.online.net:fpagny/tutorial-sdb-nextjs-terraform.git`. Note that this repository differs slightly in configuration, as it will dynamically render pages and not pre-render them in the Nextjs app build phase.

Reviewer 2: Internal link won't work. Can you upload this to a GitHub repo please?

          date: new Date("2024-01-16"),
          coverimage: "",
          author_name: "",
          author_picture: "",
          excerpt: "",
          ogimage_url: "",
          content: "",
          preview: false
        } 
      } else{
        var postRow = row
      }

      const post = {
        title: postRow.title,
        excerpt: postRow.excerpt,
        coverImage: postRow.coverimage,
        date: postRow.date.toISOString(),
        author: {
          name: postRow.author_name,
          picture: postRow.author_picture
        },
        ogImage: {
          url: postRow.ogimage_url
        },
        slug: postRow.slug,
        content: postRow.content,
        preview: false
      }
      return post;
    }

    export async function getPostBySlug(slug: string): Promise<Post> {
      const data = await client.query('SELECT * FROM posts WHERE slug=$1;',[slug])
      const post = getPostFromRow(data.rows[0])
      return post;
    }

    export async function getAllPosts(): Promise<Post[]> {
      const data = await client.query('SELECT * FROM posts;')
      const posts = data.rows.map((row) => (getPostFromRow(row)))
      return posts;
    }
    ```

> 2. edit the file located at `src/app/page.tsx` by replacing its content with the following code. This will update the main application home page which displays the posts list.

Reviewer 3: ```suggestion
2. Edit the file located at `src/app/page.tsx` by replacing its content with the following code. This will update the main application home page which displays the posts list.
```

            <Header />
            <article className="mb-32">
              <PostHeader
                title={post.title}
                coverImage={post.coverImage}
                date={post.date}
                author={post.author}
              />
              <PostBody content={post.content} />
            </article>
          </Container>
        </main>
      );
    }
    
    type Params = {
      params: {
        slug: string;
      };
    };
    
    export async function generateMetadata({ params }: Params): Promise<Metadata> {
      const post = await getPostBySlug(params.slug);
    
      if (!post) {
        return notFound();
      }
    
      const title = `${post.title} | Next.js Blog Example with ${CMS_NAME}`;
    
      return {
        openGraph: {
          title,
          images: [post.ogImage.url],
        },
      };
    }
    
    export async function generateStaticParams() {
      const posts = await getAllPosts();
    
      return posts.map((post) => ({
        slug: post.slug,
      }));
    }
    ```
  
4. Run the following command to ensure that you can run the application locally:
    ```bash
    npm run dev
    ```
   

> 5. Go to [localhost:3000](http://localhost:3000) in your browser. The images and titles of the first blog post have been updated with the content from your database.

Reviewer 2: ```suggestion
5. Go `http://localhost:3000` in your browser. The images and titles of the first blog post have been updated with the content from your database.
```

> 5. Go to [localhost:3000](http://localhost:3000) in your browser. The images and titles of the first blog post have been updated with the content from your database.

Reviewer: ```suggestion
5. Copy and paste `http://localhost:3000`  in your browser. The images and titles of the first blog post have been updated with the content from your database.
```

   
Congratulations! You can already deploy your application by building a new container version, pushing it to the **Container Registry** and deploying it to **Serverless Containers**.

> However, your application would then connect with your [user credentials](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#user), which is not a recommended security practice.
> To secure your deployment, we will now add a dedicated [IAM application](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#application), give it the minimum required permissions, and provide its credentials to your application.

Reviewer 2: ```suggestion
However, your application would then connect with your [user credentials](/identity-and-access-management/iam/concepts/#user), which is not a recommended security practice.
To secure your deployment, we will now add a dedicated [IAM application](/identity-and-access-management/iam/concepts/#application), give it the minimum required permissions, and provide its credentials to your application.
```

## Secure and redeploy your application

> 1. Run the following command to create an [IAM application](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#application) and export it as a variable: 

Reviewer 2: ```suggestion
1. Run the following command to create an [IAM application](/identity-and-access-management/iam/concepts/#application) and export it as a variable: 
```

    ```bash
    export SCW_APPLICATION_ID=$(scw iam application create name=tutorial-nextjs-app -o json | jq -r '.id')
    ```
    The `SCW_APPLICATION_ID` environment variable will store the IAM application ID so you can use it in later commands.

2. Create an [IAM policy](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#policy) giving your application rights to access the database:
    ```bash
    scw iam policy create application-id=$SCW_APPLICATION_ID name=tutorial-nextjs-policy rules.0.organization-id=$(scw config get default-organization-id) rules.0.permission-set-names.0=ServerlessSQLDatabaseFullAccess
    ```
    The permission `ServerlessSQLDatabaseFullAccess` lets your application read and write data or create new databases. You can restrict this later to fit your database use cases, using `ServerlessSQLDatabaseRead` or `ServerlessSQLDatabaseReadWrite` permissions sets.

> 3. Create an [IAM Secret Key](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#api-key):

Reviewer 2: ```suggestion
3. Create an [IAM Secret Key](/identity-and-access-management/iam/concepts/#api-key):
```

Reviewer: Should be an API key here, you can't create a secret key on its own without creating an API key. You could specify that they will only need the API key's secret key though!

    ```bash
    export SCW_APPLICATION_SECRET=$(scw iam api-key create application-id=$SCW_APPLICATION_ID -o json | jq -r '.secret_key')
    ```

4. Build your application container with the updated code:
    ```bash
    docker build -t my-nextjs-blog --build-arg PGHOST --build-arg PGDATABASE --build-arg PGPORT --build-arg PGUSER=$SCW_APPLICATION_ID --build-arg PGPASSWORD=$SCW_APPLICATION_SECRET . 
    ```
    This time build command requires environment variables because Nextjs pre-rendering will access your database at built time to load pages content.

    You can check that your container is running correctly locally with:
    ```bash
    docker run -it -p 8080:8080 -e PGHOST -e PGDATABASE -e PGPORT -e PGUSER=$SCW_APPLICATION_ID -e PGPASSWORD=$SCW_APPLICATION_SECRET my-nextjs-blog
    ```

> 5. Go to [localhots:8080](http://localhost:8080) in your browser. The blog template displays.

Reviewer 2: ```suggestion
5. Go to `http://localhost:8080` in your browser. The blog template displays.
```

6. Push your containerized application to the Container Registry:
    ```bash
    docker tag my-nextjs-blog:latest $REGISTRY_ENDPOINT/my-nextjs-blog:latest
    docker push $REGISTRY_ENDPOINT/my-nextjs-blog:latest
    ```

7. Retrieve your container ID:
    ```bash
    export CONTAINER_ID=$(scw container container list name="my-nextjs-blog" -o json | jq -r '.[0].id')
    ```

8. Redeploy your Serverless Container:
    ```bash
    scw container container deploy $CONTAINER_ID
    ```

> 9. Refresh your browser page displaying the blog. An updated version  displays. 

Reviewer 2: ```suggestion
9. Refresh your browser page displaying the blog. An updated version displays. 
```

> 9. Refresh your browser page displaying the blog. An updated version  displays. 

Reviewer 4: ```suggestion
9. Refresh your browser page displaying the blog. An updated version displays. 
```

  Congratulations, you have deployed a full serverless Next.js application!
   

> ## Going Further

Reviewer 4: Quelle est la différence entre ce "going further" ici et celui plus bas ? Les deux ont le même titre et sont au même niveau

Author: Celui-ci c'est pour la partie step-by-step avec containers, celui du bas pour la partie Terraform, ils ne sont pas 100% identiques

>   - **Edit the source code** locally, apply `scw container deploy` again, and see the **new version being live** in seconds.

Reviewer 4: ```suggestion
  - **Edit the source code** locally, apply `scw container deploy` again, and see the **new version go live** in seconds.
```

- Inspect your newly created resources in the Scaleway console:

>   - You can display your newly created **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces)

Reviewer 3: ```suggestion
  - You can display your newly created **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces)
```
```suggestion
  - You can display your **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces).
```

>   - You can display your newly created **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces)

Reviewer 3: ```suggestion
  - You can display your **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces).
```

>   - You can display your newly created **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases)

Reviewer 3: ```suggestion
  - You can display your **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases).
```

- **Fine-tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/).

## Deploy a Next.js application using Terraform templates

<Message type="note">
  You must have [Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli) installed and configured with [Scaleway Terraform Provider](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs) for this part.
</Message>

## Initialize the project

1. Run the command below to export your access key and secret key as environment variables:
    ```sh
    export SCW_ACCESS_KEY=$(scw config get access-key)
    export SCW_SECRET_KEY=$(scw config get secret-key)
    ```

2. Run the command below to make sure the environment variables are properly set:

    ```sh
    scw info 
    ```

>     This command displays your access_key and secret_key as the last two lines. The `ORIGIN` column displays `env (SCW_ACCESS_KEY)` and `env (SCW_SECRET_KEY)`, and not `default profile`.

Reviewer: ```suggestion
    This command displays your access_key and secret_key in the two last lines of the output. The `ORIGIN` column should display `env (SCW_ACCESS_KEY)` and `env (SCW_SECRET_KEY)`, and not `default profile`.
```

    
    ```bash
    KEY                      VALUE                                       ORIGIN
    (...)
    access_key               SCWF9DETBF829TAFB1TB                        env (SCW_ACCESS_KEY)
    secret_key               9a1bcf92-xxxx-xxxx-xxxx-xxxxxxxxxxxx        env (SCW_SECRET_KEY)
    ```

3. Run the following command to clone the template source code from the repository into a new folder and access it:
    ```bash

>     git clone git@gitlab.infra.online.net:fpagny/tutorial-sdb-nextjs-terraform.git my-nextjs-blog
>     && cd my-nextjs-blog

Reviewer 2: ```suggestion
    git clone git@gitlab.infra.online.net:fpagny/tutorial-sdb-nextjs-terraform.git my-nextjs-blog
    && cd my-nextjs-blog
```

Internal link won't work

    ```
      <Message type="note">
        This repository extends the blog demo application from `npx create-next-app --example blog-starter my-nextjs-blog`, and adds the required code to fetch data from a database instead of static markdown files.
      </Message>
4. Install the required dependencies:
    ```bash
    npm install
    ```
5. Confirm you can run the application locally with the following commands:
    ```bash
    npm run dev
    ```

> 6. Go to [localhost:3000](http://localhost:3000) in a browser. The blog template displays.

Reviewer 2: ```suggestion
6. Go to `http://localhost:3000` in a browser. The blog template displays.
```

### Building the image containing your Next.js app

1. Build your application container with:
    ```bash
    docker build -t my-nextjs-blog .   
    ```
    <Message type="tip">
      An error message `Error: connect ECONNREFUSED` might display when building. This is linked to Next.js prerendering stage without any database to connect to yet, but will raise any issue at runtime because all pages will be dynamically rendered. 
    </Message>

2. Run the following command to check that your container runs locally:
    ```bash
    docker run -it -p 8080:8080 my-nextjs-blog
    ```
3. Go to [localhots:8080](http://localhost:8080) in your browser. The message below displays because your application is not connected to the database yet.
    ```
    Application error: a server-side exception has occurred (see the server logs for more information)
    ```

### Pushing the image containing your Next.js app

> 1. Run the following command to create a [Container Registry namespace](https://www.scaleway.com/en/docs/containers/container-registry/concepts/#namespace) and export its endpoint as a variable:

Reviewer 2: ```suggestion
1. Run the following command to create a [Container Registry namespace]/containers/container-registry/concepts/#namespace) and export its endpoint as a variable:
```

    ```bash
    export REGISTRY_ENDPOINT=$(scw registry namespace create -o json | jq -r '.endpoint')
    ```

2. Log in to your [Container Registry](/containers/container-registry/concepts/#container-registry) from your local terminal:
    ```bash
    docker login $REGISTRY_ENDPOINT -u nologin --password-stdin <<< "$SCW_SECRET_KEY"
    ```
    The following output displays:
    ```sh
    Login Succeeded
    ```

3. Tag and push your container image to your Container Registry namespace:
    ```bash
    docker tag my-nextjs-blog:latest $REGISTRY_ENDPOINT/my-nextjs-blog:latest
    docker push $REGISTRY_ENDPOINT/my-nextjs-blog:latest
      ```
  
### Creating the Terraform configuration

1. Run the following command to create a new folder inside `my-nextjs-blog` to store your Terraform files, and access it:
    ```bash
    cd ..
    mkdir terraform-nextjs-blog
    ``` 

> 2. Access the folder you created terraform directory and create a `main.tf` terraform file:

Reviewer 3: ```suggestion
2. Access the folder in which you created the Terraform directory and create a `main.tf` Terraform file:
```

    ```bash
    cd terraform-nextjs-blog
    touch main.tf
    ``` 

    Your application folder should now be structured as follows:
    ```bash
    my-nextjs-blog/
    terraform-nextjs-blog/
        main.tf
    ```

> 9. Add the following code to your `main.tf` file:

Reviewer 4: ```suggestion
3. Add the following code to your `main.tf` file:
```

    }
      
    resource scaleway_iam_policy "db_access" {
    name = "tutorial-nextjs-policy-tf"
    description = "Gives tutorial Nextjs app access to Serverless SQL Database"
    application_id = scaleway_iam_application.app.id
    rule {
      project_ids = ["${var.DEFAULT_PROJECT_ID}"]
      permission_set_names = ["ServerlessSQLDatabaseReadWrite"]
    }
    }
      
    resource scaleway_iam_api_key "api_key" {
    application_id = scaleway_iam_application.app.id
    }
      
    resource scaleway_sdb_sql_database "database" {
    name = "tutorial-nextjs-tf"
    min_cpu = 0
    max_cpu = 8
    }
      
    output "database_connection_string" {
    // Output as an example, you can give this string to your application
    value = format("postgres://%s:%s@%s",
      scaleway_iam_application.app.id,
      scaleway_iam_api_key.api_key.secret_key,
      trimprefix(scaleway_sdb_sql_database.database.endpoint, "postgres://"),
    )
    sensitive = true
    }
      
    output "container_url" {
    // Output as an example, you can give this string to your application
    value = scaleway_container.main.domain_name
    sensitive = true
    }
    ```

> The terraform file creates several resources:

Reviewer 3: ```suggestion
The Terraform file creates several resources:
```

> - A [Serverless Containers namespace](https://www.scaleway.com/en/docs/serverless/containers/concepts/#namespace), that contains a [Serverless Container](https://www.scaleway.com/en/docs/serverless/containers/quickstart/), which hosts your Next.js application

Reviewer 2: ```suggestion
- A [Serverless Containers namespace](/serverless/containers/concepts/#namespace), that contains a [Serverless Container](/serverless/containers/quickstart/), which hosts your Next.js application
```

> - A [Serverless SQL Database](https://www.scaleway.com/en/docs/serverless/sql-databases/reference-content/serverless-sql-databases-overview/) which stores posts data

Reviewer 2: ```suggestion
- A [Serverless SQL Database](/serverless/sql-databases/reference-content/serverless-sql-databases-overview/) which stores posts data
```

- An [IAM policy](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#policy) which grants your application the right permissions

> - An [IAM application](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#application), used as a principal for your IAM policy

Reviewer 2: ```suggestion
- An [IAM application](/identity-and-access-management/iam/concepts/#application), used as a principal for your IAM policy
```

> - An [IAM secret key](https://www.scaleway.com/en/docs/identity-and-access-management/iam/concepts/#api-key) used as credentials to authenticate your application to the database

Reviewer 2: ```suggestion
- An [IAM secret key](/identity-and-access-management/iam/concepts/#api-key) used as credentials to authenticate your application to the database
```

Reviewer: Same comment as above about the API key

### Deploying the app with Terraform

1. Run the following command to initialize the Terraform working directory:
    ```bash
    terraform init
    ```

2. Add the `REGISTRY_ENDPOINT` and `DEFAULT_PROJECT_ID` environment variables to Terraform:
    ```bash
    export TF_VAR_REGISTRY_ENDPOINT=$REGISTRY_ENDPOINT
    export TF_VAR_DEFAULT_PROJECT_ID=$(scw config get default-project-id) 
    ```

> 3. Create and review Terraform the execution plan of your infrastructure:

Reviewer 3: Feels like there is something missing here, between 'and review' and 'Terraform'.

     '/assets/blog/preview/cover.jpg',
     '2020-03-16T05:35:07.322Z',
     'Joe Haddad',
     '/assets/blog/authors/joe.jpeg',
     '/assets/blog/preview/cover.jpg',
     'preview',
     'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Praesent elementum facilisis leo vel fringilla est ullamcorper eget. At imperdiet dui accumsan sit amet nulla facilities morbi tempus. Praesent elementum facilisis leo vel fringilla. Congue mauris rhoncus aenean vel. Egestas sed tempus urna et pharetra pharetra massa massa ultricies.');
   ```

4. Exit the database query interface with the following command:
    ```sql
    exit
    ```

5. Run the command below again, and open the container URL in a browser to access your application.
    ```bash
    terraform output -json
    ```

    Your serverless application is now online.

6. Once you are done, run the following command to stop all your resources:
    ```hcl
    terraform destroy
    ```

### Going further

> - **Edit the source code** locally, then:

Reviewer 4: il manque une indentation ici non ?

  - Build a new container version with the following command:

      ```bash
      docker build -t my-nextjs-blog .
      ```

  - Push the new container to **Container Registry**:

    ```bash
    docker tag my-nextjs-blog:latest $REGISTRY_ENDPOINT/my-nextjs-blog:latest 
    docker push $REGISTRY_ENDPOINT/my-nextjs-blog:latest`
    ```

  - Update and apply a new container deployment:

    ```
    export CONTAINER_ID=$(scw container container list name="my-nextjs-blog" -o json | jq -r '.[0].id')
    scw container container deploy $CONTAINER_ID
    ```

  - The new version is live a few seconds later.
  

> - Inspect your newly created resources in the Scaleway console:

Reviewer 4: indentation ici non ?

>   - You can display your newly created **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces)

Reviewer 3: ```suggestion
  - You can display your **Registry namespace** and **container image** in the [Container Registry section](https://console.scaleway.com/registry/namespaces).
```

>   - You can display your newly created **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces)

Reviewer 3: ```suggestion
  - You can display your **Serverless Containers namespace** and **container deployment** in the [Serverless Containers section](https://console.scaleway.com/containers/namespaces)
```

>   - You can display your newly created **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases)

Reviewer 3: ```suggestion
  - You can display your **Serverless SQL Database** in the [Serverless SQL Databases section](https://console.scaleway.com/serverless-db/databases).
```

> - **Fine-tune deployment options** such as autoscaling, targeted regions, and more. You can find more information by typing `scw container deploy --help` in your terminal, or by referring to the [dedicated documentation](/serverless/containers/how-to/manage-a-container/).

Reviewer 4: indentation ?

> - **Change the security configuration** for your container. It is currently public and anyone with the link can access it, but you can make it [private to require authentication](https://www.scaleway.com/en/developers/api/serverless-containers/#authentication).

Reviewer 4: indentation ?

 
## Troubleshooting

> If you encounter any issues, first check you meet all the requirements.

Reviewer 4: ```suggestion
If you happen to encounter any issues, first check that you meet all the requirements.
```

- You have installed and configured the [Scaleway CLI](https://www.scaleway.com/en/docs/developer-tools/scaleway-cli/quickstart/). Running the `scw account project get` command should return the following output:
    ```
    ID              example-7afa-4a5d-9f6c-27db072f1527
    Name            default
    OrganizationID  example-7afa-4a5d-9f6c-27db072f1527
    CreatedAt       1 year ago
    UpdatedAt       1 year ago
    Description     -
    ```
  <Message type="tip">  

>   You can also find and compare your project and organization id in the [Scaleway console settings](https://console.scaleway.com/project/settings)

Reviewer 4: ```suggestion
  You can also find and compare your project and organization ID in the [Scaleway console settings](https://console.scaleway.com/project/settings)
```

>   You can also find and compare your project and organization id in the [Scaleway console settings](https://console.scaleway.com/project/settings)

Reviewer: ```suggestion
  You can also find and compare your Project and Organization ID in the [Scaleway console settings](https://console.scaleway.com/project/settings)
```
